src/compiler/scala/tools/nsc/Reporting.scala

```diff
@@ -447,10 +453,13 @@ object Reporting {
     }
 
     final case class SourcePattern(pattern: Regex) extends MessageFilter {
-      def matches(message: Message): Boolean = {
-        // TODO: message.pos.source.file, relativize, check
-        true
-      }
+      private[this] val cache = mutable.Map.empty[SourceFile, Boolean]
+
+      def matches(message: Message): Boolean = cache.getOrElseUpdate(message.pos.source, {
+        var sourcePath = message.pos.source.file.canonicalPath.replace("\\", "/")
```

Comment:
Should we relativize paths against some root directories? Without that, builds technically will not be reproducible because absolute paths can be different on every system. I suppose there is a non-zero chance that this may cause problems - for example, a regex may accidentally match a name of some parent directory well outside of the project structure.

Reply:
I guess you're right... one root directory should be enough, no? What do you think about keeping the current behavior if no root directory is specified?

Reply:
That's exactly what silencer does.

Reply:
I wonder if the concept of root source directory could be useful elsewhere in the compiler than just for silencing warnings.

Reply:
It would definitely be useful to Scala.js, to store source positions that are not absolute.

Reply:
Does Scala.js not have any custom compiler flags?

I guess the most sensible root directory would be the project root, i.e., the path of the git checkout. `-rootdir`.

For `-Wconf`, we could also interpret `src=/...` as a full absolute regex if it starts with `/`, but that would mean repeating the project root path in every `src` filter.

Reply:
> Does Scala.js not have any custom compiler flags?

Yes, we do, but they're only useful if people use them. ;)

Reply:
Added the `-rootdir` setting in the latest commit.